File: src/core/registry.ts
import type { Provider } from "./types.js";

const registry = new Map<string, Provider>();

export function registerProvider(provider: Provider): void {
  registry.set(provider.name, provider);
}

export function unregisterProvider(name: string): boolean {
  return registry.delete(name);
}

export function getProvider(name: string): Provider | undefined {
  return registry.get(name);
}

export function listProviders(): Provider[] {
  return Array.from(registry.values());
}

export function detectProvider(key: string): Provider | undefined {
  for (const provider of registry.values()) {
    if (provider.detect(key)) return provider;
  }
  return undefined;
}

File: python/src/llm_key_validator/core/registry.py
from __future__ import annotations

from .types import Provider

_registry: dict[str, Provider] = {}


def register_provider(provider: Provider) -> None:
    _registry[provider.name] = provider


def unregister_provider(name: str) -> bool:
    return _registry.pop(name, None) is not None


def get_provider(name: str) -> Provider | None:
    return _registry.get(name)


def list_providers() -> list[Provider]:
    return list(_registry.values())


def detect_provider(key: str) -> Provider | None:
    for provider in _registry.values():
        if provider.detect(key):
            return provider
    return None
